
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="..">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.0.9">
    
    
      
        <title>Reference - Mapping mutators with inter-haplotype distance</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.0d440cfe.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.2505c338.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent="">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#scripts.utils" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Mapping mutators with inter-haplotype distance" class="md-header__button md-logo" aria-label="Mapping mutators with inter-haplotype distance" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Mapping mutators with inter-haplotype distance
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Reference
            
          </span>
        </div>
      </div>
    </div>
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Mapping mutators with inter-haplotype distance" class="md-nav__button md-logo" aria-label="Mapping mutators with inter-haplotype distance" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Mapping mutators with inter-haplotype distance
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Mapping mutator alleles with inter-haplotype distance
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Reference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Reference
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#scripts.utils" class="md-nav__link">
    scripts.utils
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.compute_colmean" class="md-nav__link">
    compute_colmean()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.compute_haplotype_distance" class="md-nav__link">
    compute_haplotype_distance()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.compute_spectra" class="md-nav__link">
    compute_spectra()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.manual_cosine_distance" class="md-nav__link">
    manual_cosine_distance()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.perform_ihd_scan" class="md-nav__link">
    perform_ihd_scan()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.perform_permutation_test" class="md-nav__link">
    perform_permutation_test()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.shuffle_spectra" class="md-nav__link">
    shuffle_spectra()
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#scripts.utils" class="md-nav__link">
    scripts.utils
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.compute_colmean" class="md-nav__link">
    compute_colmean()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.compute_haplotype_distance" class="md-nav__link">
    compute_haplotype_distance()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.compute_spectra" class="md-nav__link">
    compute_spectra()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.manual_cosine_distance" class="md-nav__link">
    manual_cosine_distance()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.perform_ihd_scan" class="md-nav__link">
    perform_ihd_scan()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.perform_permutation_test" class="md-nav__link">
    perform_permutation_test()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scripts.utils.shuffle_spectra" class="md-nav__link">
    shuffle_spectra()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



  <h1>Reference</h1>

<p>See below for descriptions of relevant method for computing IHD.</p>


<div class="doc doc-object doc-module">


<a id="scripts.utils"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">









<div class="doc doc-object doc-function">



<h2 id="scripts.utils.compute_colmean" class="doc doc-heading">
<code class="highlight language-python"><span class="n">compute_colmean</span><span class="p">(</span><span class="n">a</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Compute the mean of a 2D numpy array
on a per-column basis. Since <code>numba</code> does not
support kwargs in the <code>np.mean</code> function, it's
necessary to piece this out into its own function
so that it can be decorated with <code>numba.njit</code>'.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>a</code></td>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>A 2D numpy array of size (N, M).</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>np.ndarray: A 1D numpy array of size (M, ).</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>scripts/utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@numba</span><span class="o">.</span><span class="n">njit</span>
<span class="k">def</span> <span class="nf">compute_colmean</span><span class="p">(</span><span class="n">a</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the mean of a 2D numpy array</span>
<span class="sd">    on a per-column basis. Since `numba` does not</span>
<span class="sd">    support kwargs in the `np.mean` function, it&#39;s</span>
<span class="sd">    necessary to piece this out into its own function</span>
<span class="sd">    so that it can be decorated with `numba.njit`&#39;.</span>

<span class="sd">    Args:</span>
<span class="sd">        a (np.ndarray): A 2D numpy array of size (N, M).</span>

<span class="sd">    Returns:</span>
<span class="sd">        np.ndarray: A 1D numpy array of size (M, ).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">empty_a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="n">empty_a</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">a</span><span class="p">[:,</span> <span class="n">i</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">empty_a</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="scripts.utils.compute_haplotype_distance" class="doc doc-heading">
<code class="highlight language-python"><span class="n">compute_haplotype_distance</span><span class="p">(</span><span class="n">a_haps</span><span class="p">,</span> <span class="n">b_haps</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Compute the cosine distance between the aggregate
mutation spectrum of two collections of haplotype mutation data.
The input arrays should both be 2D numpy arrays of size (N, M), 
with rows and columns corresponding to samples and mutation types, 
respectively. This method will first aggregate the mutation spectra across 
samples to create two new 1D arrays, each of size (M, ). Then, it 
will compute the cosine distance between  those two 1D arrays. </p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>a_haps</code></td>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>2D array of size (N, M) containing the mutation </p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>b_haps</code></td>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>2D array of size (N, M) containing the mutation </p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.float64">float64</span></code>
          </td>
          <td><p>np.float64: Cosine distance between the aggregate mutation spectra </p></td>
        </tr>
        <tr>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.float64">float64</span></code>
          </td>
          <td><p>of the two haplotypes.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>scripts/utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@numba</span><span class="o">.</span><span class="n">njit</span>
<span class="k">def</span> <span class="nf">compute_haplotype_distance</span><span class="p">(</span>
    <span class="n">a_haps</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">b_haps</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the cosine distance between the aggregate</span>
<span class="sd">    mutation spectrum of two collections of haplotype mutation data.</span>
<span class="sd">    The input arrays should both be 2D numpy arrays of size (N, M), </span>
<span class="sd">    with rows and columns corresponding to samples and mutation types, </span>
<span class="sd">    respectively. This method will first aggregate the mutation spectra across </span>
<span class="sd">    samples to create two new 1D arrays, each of size (M, ). Then, it </span>
<span class="sd">    will compute the cosine distance between  those two 1D arrays. </span>

<span class="sd">    Args:</span>
<span class="sd">        a_haps (np.ndarray): 2D array of size (N, M) containing the mutation </span>
<span class="sd">        spectrum of each sample, where N is the number of samples and M is </span>
<span class="sd">        the number of mutation types.</span>
<span class="sd">        b_haps (np.ndarray): 2D array of size (N, M) containing the mutation </span>
<span class="sd">        spectrum of each sample, where N is the number of samples and M is the </span>
<span class="sd">        number of mutation types.</span>

<span class="sd">    Returns:</span>
<span class="sd">        np.float64: Cosine distance between the aggregate mutation spectra </span>
<span class="sd">        of the two haplotypes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># first, sum the spectrum arrays such that we add up the</span>
    <span class="c1"># total number of C&gt;T, C&gt;A, etc. across all samples.</span>
    <span class="n">a_hap_sums</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">a_haps</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">b_hap_sums</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">b_haps</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">dist</span> <span class="o">=</span> <span class="n">manual_cosine_distance</span><span class="p">(</span><span class="n">a_hap_sums</span><span class="p">,</span> <span class="n">b_hap_sums</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">dist</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="scripts.utils.compute_spectra" class="doc doc-heading">
<code class="highlight language-python"><span class="n">compute_spectra</span><span class="p">(</span><span class="n">singletons</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Compute the mutation spectrum of every sample in 
the input pd.DataFrame of mutation data. The input
dataframe should either contain a single entry for every mutation observed
in every sample, or the aggregate count of each mutation type
observed in each sample. The input dataframe must have at least three columns:
'Strain' (denoting the sample), 'kmer' (denoting the 3-mer context of the 
mutation -- e.g, CCT&gt;CAT), and 'count' (denoting the number of times
a mutation of type <code>kmer</code> was observed in <code>Strain</code>). If the dataframe contains
a single entry for every mutation observed in every sample, then the 'count'
column should contain a value of 1 in every row.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>singletons</code></td>
          <td>
                <code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
          </td>
          <td><p>Pandas dataframe containing information </p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>at</code></td>
          <td>
                <code>least three columns</code>
          </td>
          <td><p>Strain (denoting the sample), kmer (denoting </p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>k</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>k-mer size of mutations (e.g., k=3 means that we </p></td>
          <td>
                <code>1</code>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.Iterable">Iterable</span>[tuple[<span title="typing.List">List</span>[str], <span title="typing.List">List</span>[str], <span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span>]]</code>
          </td>
          <td><p>Tuple[List[str], List[str], np.ndarray,]: A tuple of three objects:</p></td>
        </tr>
        <tr>
          <td>
                <code><span title="typing.Iterable">Iterable</span>[tuple[<span title="typing.List">List</span>[str], <span title="typing.List">List</span>[str], <span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span>]]</code>
          </td>
          <td><ol>
<li>A list of samples in the dataframe (which are indexed in the same</li>
</ol></td>
        </tr>
        <tr>
          <td>
                <code><span title="typing.Iterable">Iterable</span>[tuple[<span title="typing.List">List</span>[str], <span title="typing.List">List</span>[str], <span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span>]]</code>
          </td>
          <td><p>order as the rows of the 2D <code>spectra</code> array).</p></td>
        </tr>
        <tr>
          <td>
                <code><span title="typing.Iterable">Iterable</span>[tuple[<span title="typing.List">List</span>[str], <span title="typing.List">List</span>[str], <span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span>]]</code>
          </td>
          <td><ol>
<li>A list of the unique mutation types observed in the dataframe.</li>
</ol></td>
        </tr>
        <tr>
          <td>
                <code><span title="typing.Iterable">Iterable</span>[tuple[<span title="typing.List">List</span>[str], <span title="typing.List">List</span>[str], <span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span>]]</code>
          </td>
          <td><ol>
<li>A 2D numpy array of mutation spectra in all genotyped</li>
</ol></td>
        </tr>
        <tr>
          <td>
                <code><span title="typing.Iterable">Iterable</span>[tuple[<span title="typing.List">List</span>[str], <span title="typing.List">List</span>[str], <span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span>]]</code>
          </td>
          <td><p>samples, of size (N, M) where N is the number of samples and M is the number of</p></td>
        </tr>
        <tr>
          <td>
                <code><span title="typing.Iterable">Iterable</span>[tuple[<span title="typing.List">List</span>[str], <span title="typing.List">List</span>[str], <span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span>]]</code>
          </td>
          <td><p>mutation types.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>scripts/utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compute_spectra</span><span class="p">(</span>
    <span class="n">singletons</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the mutation spectrum of every sample in </span>
<span class="sd">    the input pd.DataFrame of mutation data. The input</span>
<span class="sd">    dataframe should either contain a single entry for every mutation observed</span>
<span class="sd">    in every sample, or the aggregate count of each mutation type</span>
<span class="sd">    observed in each sample. The input dataframe must have at least three columns:</span>
<span class="sd">    &#39;Strain&#39; (denoting the sample), &#39;kmer&#39; (denoting the 3-mer context of the </span>
<span class="sd">    mutation -- e.g, CCT&gt;CAT), and &#39;count&#39; (denoting the number of times</span>
<span class="sd">    a mutation of type `kmer` was observed in `Strain`). If the dataframe contains</span>
<span class="sd">    a single entry for every mutation observed in every sample, then the &#39;count&#39;</span>
<span class="sd">    column should contain a value of 1 in every row.</span>

<span class="sd">    Args:</span>
<span class="sd">        singletons (pd.DataFrame): Pandas dataframe containing information </span>
<span class="sd">        about the mutations observed in each sample. The dataframe must have </span>
<span class="sd">        at least three columns: Strain (denoting the sample), kmer (denoting </span>
<span class="sd">        the 3-mer context of the mutation), and `count` (denoting the number </span>
<span class="sd">        of times a mutation of type `kmer` was observed in `Strain`).</span>
<span class="sd">        k (int, optional): k-mer size of mutations (e.g., k=3 means that we </span>
<span class="sd">        will treat mutations as NXN NYN, where Ns represent the nucleotide </span>
<span class="sd">        contexts on either side of the mutation, whereas k=1 means we will </span>
<span class="sd">        treat them as X Y). Defaults to 1.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple[List[str], List[str], np.ndarray,]: A tuple of three objects:</span>
<span class="sd">        1. A list of samples in the dataframe (which are indexed in the same</span>
<span class="sd">        order as the rows of the 2D `spectra` array).</span>
<span class="sd">        2. A list of the unique mutation types observed in the dataframe.</span>
<span class="sd">        3. A 2D numpy array of mutation spectra in all genotyped</span>
<span class="sd">        samples, of size (N, M) where N is the number of samples and M is the number of</span>
<span class="sd">        mutation types.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># compute 3-mer spectra</span>
    <span class="n">hap_spectra_agg</span> <span class="o">=</span> <span class="n">singletons</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Strain&#39;</span><span class="p">,</span> <span class="s1">&#39;kmer&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span>
        <span class="s1">&#39;count&#39;</span><span class="p">:</span> <span class="nb">sum</span>
    <span class="p">})</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>  <span class="c1">#.rename(columns={0: &#39;count&#39;})</span>
    <span class="c1"># if 1-mer spectra are desired, compute that</span>
    <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="c1"># add base mutation type</span>
        <span class="n">hap_spectra_agg</span><span class="p">[</span><span class="s1">&#39;base_mut&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hap_spectra_agg</span><span class="p">[</span><span class="s1">&#39;kmer&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">k</span><span class="p">:</span> <span class="s2">&quot;&gt;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">k</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">k</span><span class="p">[</span><span class="mi">5</span><span class="p">]]))</span>
        <span class="n">hap_spectra_agg</span> <span class="o">=</span> <span class="n">hap_spectra_agg</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Strain&#39;</span><span class="p">,</span> <span class="s1">&#39;base_mut&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span>
            <span class="s1">&#39;count&#39;</span><span class="p">:</span>
            <span class="nb">sum</span>
        <span class="p">})</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="c1"># get spectra as per-haplotype vectors of mutation counts</span>
    <span class="n">mut_col</span> <span class="o">=</span> <span class="s2">&quot;base_mut&quot;</span> <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="s2">&quot;kmer&quot;</span>
    <span class="n">spectra</span> <span class="o">=</span> <span class="n">hap_spectra_agg</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="s2">&quot;Strain&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">mut_col</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">samples</span><span class="p">,</span> <span class="n">mutations</span><span class="p">,</span> <span class="n">spectra</span> <span class="o">=</span> <span class="n">spectra</span><span class="p">[</span><span class="s1">&#39;Strain&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="p">[</span>
        <span class="n">el</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">el</span> <span class="ow">in</span> <span class="n">spectra</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="p">],</span> <span class="n">spectra</span><span class="o">.</span><span class="n">values</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">:]</span>

    <span class="k">return</span> <span class="n">samples</span><span class="p">,</span> <span class="n">mutations</span><span class="p">,</span> <span class="n">spectra</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="scripts.utils.manual_cosine_distance" class="doc doc-heading">
<code class="highlight language-python"><span class="n">manual_cosine_distance</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Compute the cosine distance between
two 1D numpy arrays. Although methods to compute
cosine similarity and distance exist in <code>scipy</code> and
<code>sklearn</code>, they are not <code>numba.njit</code>'able.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>a</code></td>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>A 1D numpy array of size (N, ).</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>b</code></td>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>A 1D numpy array of size (N, ).</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.float64">float64</span></code>
          </td>
          <td><p>np.float64: Cosine distance between a and b.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>scripts/utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@numba</span><span class="o">.</span><span class="n">njit</span>
<span class="k">def</span> <span class="nf">manual_cosine_distance</span><span class="p">(</span>
    <span class="n">a</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">b</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the cosine distance between</span>
<span class="sd">    two 1D numpy arrays. Although methods to compute</span>
<span class="sd">    cosine similarity and distance exist in `scipy` and</span>
<span class="sd">    `sklearn`, they are not `numba.njit`&#39;able.</span>

<span class="sd">    Args:</span>
<span class="sd">        a (np.ndarray): A 1D numpy array of size (N, ).</span>
<span class="sd">        b (np.ndarray): A 1D numpy array of size (N, ).</span>

<span class="sd">    Returns:</span>
<span class="sd">        np.float64: Cosine distance between a and b.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">dot</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">b</span><span class="p">)</span>
    <span class="n">a_sumsq</span><span class="p">,</span> <span class="n">b_sumsq</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">square</span><span class="p">(</span><span class="n">a</span><span class="p">)),</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">square</span><span class="p">(</span><span class="n">b</span><span class="p">))</span>
    <span class="n">a_norm</span><span class="p">,</span> <span class="n">b_norm</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">a_sumsq</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">b_sumsq</span><span class="p">)</span>
    <span class="n">cossim</span> <span class="o">=</span> <span class="n">dot</span> <span class="o">/</span> <span class="p">(</span><span class="n">a_norm</span> <span class="o">*</span> <span class="n">b_norm</span><span class="p">)</span>
    <span class="k">return</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">cossim</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="scripts.utils.perform_ihd_scan" class="doc doc-heading">
<code class="highlight language-python"><span class="n">perform_ihd_scan</span><span class="p">(</span><span class="n">spectra</span><span class="p">,</span> <span class="n">genotype_matrix</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Iterate over every genotyped marker in the <code>genotype_matrix</code>, 
divide the haplotypes into two groups based on sample genotypes at the 
marker, and compute the distance between the aggregate mutation spectra
of each group. Return a list of cosine distances of length (G, ), where
G is the number of genotyped sites.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>spectra</code></td>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>A 2D numpy array of mutation spectra in </p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>genotype_matrix</code></td>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>A 2D numpy array of genotypes at </p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="numpy">np</span>.<span title="numpy.float64">float64</span>]</code>
          </td>
          <td><p>List[np.float64]: List of inter-haplotype cosine distances at every </p></td>
        </tr>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="numpy">np</span>.<span title="numpy.float64">float64</span>]</code>
          </td>
          <td><p>marker.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>scripts/utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@numba</span><span class="o">.</span><span class="n">njit</span><span class="p">()</span>
<span class="k">def</span> <span class="nf">perform_ihd_scan</span><span class="p">(</span>
    <span class="n">spectra</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">genotype_matrix</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Iterate over every genotyped marker in the `genotype_matrix`, </span>
<span class="sd">    divide the haplotypes into two groups based on sample genotypes at the </span>
<span class="sd">    marker, and compute the distance between the aggregate mutation spectra</span>
<span class="sd">    of each group. Return a list of cosine distances of length (G, ), where</span>
<span class="sd">    G is the number of genotyped sites.</span>

<span class="sd">    Args:</span>
<span class="sd">        spectra (np.ndarray): A 2D numpy array of mutation spectra in </span>
<span class="sd">        all genotyped samples, of size (N, M) where N is the number of </span>
<span class="sd">        samples and M is the number of mutation types.</span>
<span class="sd">        genotype_matrix (np.ndarray): A 2D numpy array of genotypes at </span>
<span class="sd">        every genotyped marker, of size (G, N), where G is the number </span>
<span class="sd">        of genotyped sites and N is the number of samples.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[np.float64]: List of inter-haplotype cosine distances at every </span>
<span class="sd">        marker.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># store distances at each marker</span>
    <span class="n">focal_dist</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># loop over every site in the genotype matrix</span>
    <span class="k">for</span> <span class="n">ni</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">genotype_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">a_hap_idxs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">genotype_matrix</span><span class="p">[</span><span class="n">ni</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">b_hap_idxs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">genotype_matrix</span><span class="p">[</span><span class="n">ni</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">a_spectra</span><span class="p">,</span> <span class="n">b_spectra</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">spectra</span><span class="p">[</span><span class="n">a_hap_idxs</span><span class="p">],</span>
            <span class="n">spectra</span><span class="p">[</span><span class="n">b_hap_idxs</span><span class="p">],</span>
        <span class="p">)</span>

        <span class="n">cur_dist</span> <span class="o">=</span> <span class="n">compute_haplotype_distance</span><span class="p">(</span><span class="n">a_spectra</span><span class="p">,</span> <span class="n">b_spectra</span><span class="p">)</span>

        <span class="n">focal_dist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_dist</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">focal_dist</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="scripts.utils.perform_permutation_test" class="doc doc-heading">
<code class="highlight language-python"><span class="n">perform_permutation_test</span><span class="p">(</span><span class="n">spectra</span><span class="p">,</span> <span class="n">genotype_matrix</span><span class="p">,</span> <span class="n">n_permutations</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Conduct a permutation test to assess the significance of 
any observed IHD peaks. In each of the <code>n_permutations</code> trials, 
do the following: 1. create a shuffled version of the input mutation <code>spectra</code>, so that
sample names/indices no longer correspond to the appropriate mutation
spectrum. 2. run an IHD scan by computing the cosine distance between
the aggregate mutation spectrum of samples with either genotype
at every marker in the <code>genotype_matrix</code>. 3. store the maximum cosine distance encountered at any marker.
Then, return a list of the maximum cosine distances encountered in each of
the trials.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>spectra</code></td>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>A 2D numpy array of mutation spectra in all </p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>genotype_matrix</code></td>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>A 2D numpy array of genotypes at every </p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>n_permutations</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>Number of permutations to perform </p></td>
          <td>
                <code>1000</code>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="numpy">np</span>.<span title="numpy.float64">float64</span>]</code>
          </td>
          <td><p>List[np.float64]: List of length <code>n_permutations</code>, containing the </p></td>
        </tr>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="numpy">np</span>.<span title="numpy.float64">float64</span>]</code>
          </td>
          <td><p>maximum cosine distance encountered in each permutation.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>scripts/utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@numba</span><span class="o">.</span><span class="n">njit</span><span class="p">()</span>
<span class="k">def</span> <span class="nf">perform_permutation_test</span><span class="p">(</span>
    <span class="n">spectra</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">genotype_matrix</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">n_permutations</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1_000</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Conduct a permutation test to assess the significance of </span>
<span class="sd">    any observed IHD peaks. In each of the `n_permutations` trials, </span>
<span class="sd">    do the following: 1. create a shuffled version of the input mutation `spectra`, so that</span>
<span class="sd">    sample names/indices no longer correspond to the appropriate mutation</span>
<span class="sd">    spectrum. 2. run an IHD scan by computing the cosine distance between</span>
<span class="sd">    the aggregate mutation spectrum of samples with either genotype</span>
<span class="sd">    at every marker in the `genotype_matrix`. 3. store the maximum cosine distance encountered at any marker.</span>
<span class="sd">    Then, return a list of the maximum cosine distances encountered in each of</span>
<span class="sd">    the trials.</span>

<span class="sd">    Args:</span>
<span class="sd">        spectra (np.ndarray): A 2D numpy array of mutation spectra in all </span>
<span class="sd">        genotyped samples, of size (N, M) where N is the number of samples </span>
<span class="sd">        and M is the number of mutation types.</span>
<span class="sd">        genotype_matrix (np.ndarray): A 2D numpy array of genotypes at every </span>
<span class="sd">        genotyped marker, of size (G, N), where G is the number of genotyped </span>
<span class="sd">        sites and N is the number of samples.</span>
<span class="sd">        n_permutations (int, optional): Number of permutations to perform </span>
<span class="sd">        (i.e., number of times to shuffle the spectra and compute IHDs at </span>
<span class="sd">        each marker). Defaults to 1_000.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[np.float64]: List of length `n_permutations`, containing the </span>
<span class="sd">        maximum cosine distance encountered in each permutation.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># store max cosdist encountered in each permutation</span>
    <span class="n">max_distances</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float16</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">pi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_permutations</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">pi</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">pi</span> <span class="o">%</span> <span class="mi">100</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">pi</span><span class="p">)</span>
        <span class="c1"># shuffle the mutation spectra by row</span>
        <span class="n">shuffled_spectra</span> <span class="o">=</span> <span class="n">shuffle_spectra</span><span class="p">(</span><span class="n">spectra</span><span class="p">)</span>
        <span class="c1"># compute the cosine distances at each marker</span>
        <span class="n">perm_distances</span> <span class="o">=</span> <span class="n">perform_ihd_scan</span><span class="p">(</span>
            <span class="n">shuffled_spectra</span><span class="p">,</span>
            <span class="n">genotype_matrix</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">max_distances</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">perm_distances</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">max_distances</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="scripts.utils.shuffle_spectra" class="doc doc-heading">
<code class="highlight language-python"><span class="n">shuffle_spectra</span><span class="p">(</span><span class="n">spectra</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Randomly shuffle the rows of a 2D numpy array of 
mutation spectrum data of size (N, M), where N is the number
of samples and M is the number of mutation types. Shuffled array
is returned such that sample mutation spectra no longer correspond to the
appropriate sample indices into the rows of the array.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>spectra</code></td>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>2D numpy array of mutation spectrum data of </p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="numpy">np</span>.<span title="numpy.ndarray">ndarray</span></code>
          </td>
          <td><p>np.ndarray: The input array, but with shuffled rows.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>scripts/utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@numba</span><span class="o">.</span><span class="n">njit</span>
<span class="k">def</span> <span class="nf">shuffle_spectra</span><span class="p">(</span><span class="n">spectra</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Randomly shuffle the rows of a 2D numpy array of </span>
<span class="sd">    mutation spectrum data of size (N, M), where N is the number</span>
<span class="sd">    of samples and M is the number of mutation types. Shuffled array</span>
<span class="sd">    is returned such that sample mutation spectra no longer correspond to the</span>
<span class="sd">    appropriate sample indices into the rows of the array.</span>

<span class="sd">    Args:</span>
<span class="sd">        spectra (np.ndarray): 2D numpy array of mutation spectrum data of </span>
<span class="sd">        shape (N, M), where N is the number of samples and M is the number </span>
<span class="sd">        of mutation types.</span>

<span class="sd">    Returns:</span>
<span class="sd">        np.ndarray: The input array, but with shuffled rows.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">idxs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">spectra</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="c1"># shuffle the spectra so that sample idxs no longer</span>
    <span class="c1"># correspond to the appropriate spectrum</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">idxs</span><span class="p">)</span>
    <span class="n">shuffled_spectra</span> <span class="o">=</span> <span class="n">spectra</span><span class="p">[</span><span class="n">idxs</span><span class="p">,</span> <span class="p">:]</span>
    <span class="k">return</span> <span class="n">shuffled_spectra</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.db81ec45.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.a00a7c5e.min.js"></script>
      
    
  </body>
</html>